src/main/java/com/sourcegraph/webhook/EventSerializer.java

```diff
@@ -0,0 +1,103 @@
+package com.sourcegraph.webhook;
+
+import com.atlassian.bitbucket.event.ApplicationEvent;
+import com.atlassian.bitbucket.event.pull.*;
+import com.atlassian.bitbucket.json.JsonRenderer;
+import com.atlassian.plugin.spring.scanner.annotation.imports.ComponentImport;
+import com.google.gson.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.text.SimpleDateFormat;
+import java.util.HashMap;
+import java.util.Map;
+
+@Component
+public class EventSerializer {
+    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
+    private static Map<Class<?>, Adapter> adapters = new HashMap<>();
+    private static JsonRenderer renderer;
+
+    private JsonObject payload;
+
+    private static JsonElement render(Object o) {
+        String raw = renderer.render(o, new HashMap<>());
+        return raw == null ? null : new JsonParser().parse(raw);
+    }
+
+    private static Adapter<PullRequestUpdatedEvent> PullRequestUpdatedEventAdapter = (element, event) -> {
```

Comment:
I considered the option of using reflection to serialize these events, but I decided against it. The reflection approach would require unnecessary code complexity (and even more to allow flexibility in the JSON output). This isn't worth it for the few number of supported events. This decision might be revisited if we support hooking onto more events.

Ideally, the Bitbucket API would expose something to serialize these events, but it doesn't :(

Reply:
The Bitbucket Webhook plugin I linked to in the RFC seems to have a way to solve [this problem generically](https://sourcegraph.com/github.com/Eernie/bitbucket-webhooks-plugin/-/blob/src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java#L308). What's in the way for us to use a similar approach?

Reply:
JSON serializers like `gson` and `jackson` (the one they use) don't work on the native bitbucket event objects. Their approach was to create [their own classes](https://sourcegraph.com/github.com/Eernie/bitbucket-webhooks-plugin/-/tree/src/main/java/nl/topicus/bitbucket/model) that mirror each native bitbucket event objects. So whenever they wanted to serialize, they would convert the native objects to one of their classes and run `jackson` on that.

I wanted to avoid creating all these extra classes, so in a way, I defined the class formats in this file through the various adapters. I have one more idea that I'm going to try.